test/core/client_channel/lb_policy/round_robin_test.cc

```diff
@@ -0,0 +1,251 @@
+        {MakeAddress(UriForSubchannel(subchannel)), ChannelArgs()});
+    return it->second.ConnectionRequested();
+  }
+
+  std::string UriForSubchannel(size_t index) {
+    std::stringstream out;
+    out << "ipv4:127.0.0.1:44" << (index + 1);
+    return out.str();
+  }
+};
+
+TEST_F(RoundRobinTest, SingleChannel) {
+  auto policy = RoundRobinPolicy(1);
+  // LB policy should have reported CONNECTING state.
+  auto picker = ExpectState(GRPC_CHANNEL_CONNECTING);
+  ExpectPickQueued(picker.get());
+  // LB policy should have requested a connection on this subchannel.
+  EXPECT_TRUE(ConnectionRequested(0));
+
+  SetSubchannelConnectivityState(0, GRPC_CHANNEL_CONNECTING);
+  picker = ExpectState(GRPC_CHANNEL_CONNECTING);
+  ExpectPickQueued(picker.get());
+
+  // Channel is ready
+  SetSubchannelConnectivityState(0, GRPC_CHANNEL_READY);
+  picker = ExpectState(GRPC_CHANNEL_READY);
+  // Picker should return the same subchannel repeatedly.
+  for (size_t i = 0; i < 3; ++i) {
+    ExpectPickComplete(picker.get(), UriForSubchannel(0));
+  }
+  ExpectNoStateChange();
+
+  // There's a failure
+  SetSubchannelConnectivityState(0, GRPC_CHANNEL_TRANSIENT_FAILURE);
+  ExpectReresolutionRequest();
+  picker = ExpectState(GRPC_CHANNEL_TRANSIENT_FAILURE);
+  ExpectPickFail(picker.get());
+
+  // ... and a recovery!
+  SetSubchannelConnectivityState(0, GRPC_CHANNEL_READY);
+  picker = ExpectState(GRPC_CHANNEL_READY);
+  ExpectPickComplete(picker.get(), UriForSubchannel(0));
+
+  ExpectNoStateChange();
+}
+
+TEST_F(RoundRobinTest, ThreeSubchannels) {
+  auto policy = RoundRobinPolicy(3);
+  // LB policy should have reported CONNECTING state.
+  auto picker = ExpectState(GRPC_CHANNEL_CONNECTING);
+  ExpectPickQueued(picker.get());
+  // LB policy should have requested a connection on this subchannel.
+  EXPECT_TRUE(ConnectionRequested(0));
+
+  for (size_t i = 0; i < 3; ++i) {
+    SetSubchannelConnectivityState(i, GRPC_CHANNEL_CONNECTING);
+    picker = ExpectState(GRPC_CHANNEL_CONNECTING);
+  }
+  ExpectPickQueued(picker.get());
+
+  // Only one channel is ready
+  SetSubchannelConnectivityState(0, GRPC_CHANNEL_READY);
+  picker = ExpectState(GRPC_CHANNEL_READY);
+  // Picker should return the same subchannel repeatedly.
+  for (size_t i = 0; i < 3; ++i) {
+    ExpectPickComplete(picker.get(), UriForSubchannel(0));
+  }
+  ExpectNoStateChange();
+
+  // All channels ready
+  SetSubchannelConnectivityState(1, GRPC_CHANNEL_READY);
+  SetSubchannelConnectivityState(2, GRPC_CHANNEL_READY);
+  ExpectState(GRPC_CHANNEL_READY);
+  picker = ExpectState(GRPC_CHANNEL_READY);
+
+  std::unordered_set<std::string> reportedUris;
+  // Picker should return each address once, we do not care about the order.
+  for (size_t i = 0; i < 20; ++i) {
+    absl::optional<std::string> address;
+    GetPickedAddress(picker.get(), address);
+    EXPECT_TRUE(address.has_value());
+    reportedUris.insert(*address);
+  }
+  EXPECT_EQ(reportedUris.size(), 3);
```

Comment:
This can be:

```
EXPECT_THAT(reportedUris, ::testing::UnorderedElementsAre(
    UriForSubchannel(0), UriForSubchannel(1), UriForSubchannel(2)));
```